### PR TITLE
[webgpu] Fix matmul_small_output program assert error

### PR DESCRIPTION
**Pull request by haoyunfeix**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5646)

### CODE REVIEW

**Comment by haoyunfeix** [comment]
@qjia7 @axinging @xhcao @gyagp, PTAL

**Comment by xhcao** [comment]
`aShape` shape is [batch, M, K], `bShape` shape is [batch, K, N], This program is applied when M and N are small. I think there is no index issue. Are there some failed cases?

**Comment by haoyunfeix** [comment]
> `aShape` shape is [batch, M, K], `bShape` shape is [batch, K, N], This program is applied when M and N are small. I think there is no index issue. Are there some failed cases?

Yes, I met error in USE model demo : `This program can be only used when A width is small.`
Firstly I thought it is a typo error, because https://github.com/tensorflow/tfjs/blob/master/tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts#L114 checked aShape not a3dShape, so if aShape = [1, b, c, d] then a3dShape =  [b, c, d].

According to your explanation, do you think we should move a3dShape|b3dShape check from https://github.com/tensorflow/tfjs/blob/master/tfjs-backend-webgpu/src/kernels/matmul_small_output_size_webgpu.ts#L224 to `BatchMatMul_impl.ts`? As we should never refuse to execute op in a program we defined.

File: tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts [hunk 1]
```diff
@@ -110,7 +110,7 @@ export function batchMatMulImpl({
   // smaller than each of the two input sizes. For example, if input sizes are
   // [12, 2048] and [2048, 1024], the output size is [12, 1024], which is
   // relatively small compared to input sizes.
-  if (!transposeA && !transposeB &&
+  if (!transposeA && !transposeB && (a3dShape[1] <= 16 || b3dShape[2] <= 16) &&
       ((a.shape[1] <= 16 &&
         (b.shape[2] <= 512 || b.shape[1] >= 2 * b.shape[2])) ||
        (b.shape[2] <= 16 &&
```
Comment (xhcao) [inline tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts:116]: Change all a.shape and b.shape to a3dShape and b3dShape as below,
if (!transposeA && !transposeB &&
      ((a3dShape[1] <= 16 &&
        (b3dShape[2] <= 512 || b3dShape[1] >= 2 * b3dShape[2])) ||
       (b3dShape[2] <= 16 &&
        (a3dShape[1] <= 512 || a3dShape[2] >= 2 * a3dShape[1]))))

**Comment by xhcao** [comment]
LGTM

File: tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts [hunk 2]
```diff
@@ -110,10 +110,10 @@ export function batchMatMulImpl({
   // [12, 2048] and [2048, 1024], the output size is [12, 1024], which is
   // relatively small compared to input sizes.
   if (!transposeA && !transposeB &&
-      ((a.shape[1] <= 16 &&
-        (b.shape[2] <= 512 || b.shape[1] >= 2 * b.shape[2])) ||
-       (b.shape[2] <= 16 &&
-        (a.shape[1] <= 512 || a.shape[2] >= 2 * a.shape[1])))) {
+      ((a3dShape[1] <= 16 &&
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts:113]: Please also change the checking of `useVec4`. And unify to use `outerShapeA` `outerShapeB` `innerShapeA[B]` for readability.

Comment (haoyunfeix) [reply tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts]: Done. PTAL.

**Review by qjia7 — APPROVED** [review]
LGTM, thanks.

cc @lina128 @pyu10055

**Pull request merged**